File: 278.first-bad-version.cpp
#include <bits/stdc++.h>
using namespace std;
/*
 * @lc app=leetcode id=278 lang=cpp
 *
 * [278] First Bad Version
 */
bool isBadVersion(int version);
// @lc code=start
// The API isBadVersion is defined for you.
// bool isBadVersion(int version);

class Solution
{
public:
    int firstBadVersion(int n)
    {
        int l = 1, r = n, b = n, m;

        while (r >= l)
        {
            m = l + (r - l) / 2;

            if (isBadVersion(m))
            {
                r = m - 1;
                b = m;
            }
            else
            {
                l = m + 1;
            }
        }
        return b;
    }
};
// @lc code=end

File: Java/Grind75/Week2/278.first-bad-version.java
package Grind75.Week2;

/*
 * @lc app=leetcode id=278 lang=java
 *
 * [278] First Bad Version
 */
class VersionControl {
    boolean isBadVersion(int version) {
        return false;
    }
}
// @lc code=start
/*
 * The isBadVersion API is defined in the parent class VersionControl.
 * boolean isBadVersion(int version);
 */

class Solution extends VersionControl {
    public int firstBadVersion(int n) {
        int l = 1, r = n;
        while (l < r) {
            int m = (r - l) / 2 + l;
            if (isBadVersion(m))
                r = m;
            else
                l = m + 1;
        }
        return l;
    }
}
// @lc code=end
